Bottom: 600fcef64230102ed3ce5979f8f8edf258d760c0
Top:    747991009b9642ea8b3e2ca288f23b55b22d602d
Author: Jon Smirl <jonsmirl@gmail.com>
Date:   2014-08-16 16:04:57 -0400

Refresh of pwm

---

diff --git a/arch/arm/boot/dts/sun7i-a20-cubietruck.dts b/arch/arm/boot/dts/sun7i-a20-cubietruck.dts
index 186f862..def4394 100644
--- a/arch/arm/boot/dts/sun7i-a20-cubietruck.dts
+++ b/arch/arm/boot/dts/sun7i-a20-cubietruck.dts
@@ -157,7 +157,7 @@
 			sgtl5000: sgtl5000@a {
 				compatible = "fsl,sgtl5000";
 				reg = <0x0a>;
-				pwms = <&pwm 0 100 0>;
+				pwms = <&pwm 0 50 0>;
 
 				#sound-dai-cells = <0>;
 				VDDA-supply = <&reg_vcc3v3>;
diff --git a/drivers/pwm/pwm-sunxi.c b/drivers/pwm/pwm-sunxi.c
index 240d01e..58880e1 100644
--- a/drivers/pwm/pwm-sunxi.c
+++ b/drivers/pwm/pwm-sunxi.c
@@ -294,6 +294,21 @@ static int sunxi_pwm_config(struct pwm_chip *chip, struct pwm_device *pwm,
 	unsigned int prescale, entire_cycles, active_cycles, reg_val;
 
 	printk("JDS - sunxi_pwm_config duty %d period %d\n", duty_ns, period_ns);
+	if (period_ns <= 50) {
+		// Just enable the OSC24 clock bypass 
+		switch (pwm->hwpwm) {
+		case 0:
+			regmap_update_bits(priv->regmap, SUNXI_PWM_CTRL_REG,
+					SUNXI_PWMCTL_PWM0_BYPASS_MASK, SUNXI_PWMCTL_PWM0_BYPASS);
+			break;
+		case 1:
+			regmap_update_bits(priv->regmap, SUNXI_PWM_CTRL_REG,
+					SUNXI_PWMCTL_PWM1_BYPASS_MASK, SUNXI_PWMCTL_PWM1_BYPASS);
+			break;
+		}
+		return 0;
+	}
+
 	prescale = pwm_get_best_prescale(period_ns);
 	if (prescale < 0)
 		return prescale;
@@ -330,12 +345,9 @@ static int sunxi_pwm_enable(struct pwm_chip *chip, struct pwm_device *pwm)
 	
 	switch (pwm->hwpwm) {
 	case 0:
-		regmap_write(priv->regmap, SUNXI_PWM_CTRL_REG, SUNXI_PWMCTL_PWM0_GATE | SUNXI_PWMCTL_PWM0_EN | SUNXI_PWMCTL_PWM0_BYPASS);
-#ifdef JDS
 		regmap_update_bits(priv->regmap, SUNXI_PWM_CTRL_REG,
 			SUNXI_PWMCTL_PWM0_GATE_MASK | SUNXI_PWMCTL_PWM0_EN_MASK,
 			SUNXI_PWMCTL_PWM0_GATE | SUNXI_PWMCTL_PWM0_EN);
-#endif
 		break;
 	case 1:
 		regmap_update_bits(priv->regmap, SUNXI_PWM_CTRL_REG,
